File: 01-clases-objetos/java/Demo01.java
/*
 * DEMO 01 - CLASES Y OBJETOS
 *
 * Compilar y ejecutar:
 *     javac *.java
 *     java Demo01
 *
 * Idea clave: una CLASE es el molde. Un OBJETO (o instancia) es cada
 * galleta que sale de ese molde. El molde no se come; las galletas si.
 *
 * Esta clase SI tiene main(): es la que se ejecuta.
 * No necesita "import" para ver a Perro porque estan en la misma carpeta.
 */
public class Demo01 {
    public static void main(String[] args) {

        // CREAR INSTANCIAS: cada "new" fabrica un objeto nuevo e independiente
        Perro firulais = new Perro("Firulais", "Mestizo", 3);
        Perro laika = new Perro("Laika", "Husky", 5);

        firulais.presentarse();
        laika.presentarse();

        System.out.println("---");

        firulais.ladrar();
        firulais.cumplirAnios();

        // Laika NO cambio: son objetos distintos, con memoria distinta
        laika.presentarse();

        System.out.println("---");

        // Los atributos se leen y se escriben con el punto
        System.out.println("La raza de laika es: " + laika.raza);
        laika.raza = "Husky Siberiano";
        laika.presentarse();
    }
}

File: 01-clases-objetos/python/demo01.py
"""
DEMO 01 - CLASES Y OBJETOS
Ejecutar:  python3 demo01.py

Mismo programa que Demo01.java, linea por linea.

EQUIVALENCIAS CON JAVA
  (nada, estan en la misma carpeta)  ->  from perro import Perro
  class Perro { }                    ->  class Perro:
  Perro(String n, ...) {}            ->  def __init__(self, n, ...):
  this.nombre                        ->  self.nombre
  new Perro("Firulais", ...)         ->  Perro("Firulais", ...)   (sin "new")
  void ladrar() { }                  ->  def ladrar(self):        (self explicito)

OJO CON EL IMPORT: se lee "del archivo perro.py, traeme la clase Perro".
  from perro import Perro
       ^^^^^             archivo (minuscula, sin .py)
              ^^^^^^^^^^ clase   (CamelCase)

En Java no hace falta importar nada porque las dos clases estan en la
misma carpeta (el mismo "paquete"). En Python el import SIEMPRE es necesario.
"""

from perro import Perro


# Este "if" es el equivalente de public static void main(String[] args)
if __name__ == "__main__":

    # CREAR INSTANCIAS: sin "new", solo se llama a la clase
    firulais = Perro("Firulais", "Mestizo", 3)
    laika = Perro("Laika", "Husky", 5)

    firulais.presentarse()
    laika.presentarse()

    print("---")

    firulais.ladrar()
    firulais.cumplir_anios()

    # Laika NO cambio: son objetos distintos
    laika.presentarse()

    print("---")

    # Los atributos se leen y se escriben con el punto, igual que en Java
    print(f"La raza de laika es: {laika.raza}")
    laika.raza = "Husky Siberiano"
    laika.presentarse()
